easybuild/framework/easyconfig/easyconfig.py

```diff
@@ -195,44 +217,54 @@ def get_toolchain_hierarchy(parent_toolchain):
-                subtoolchain_version = ''
+        # only retain candidates that match subtoolchain names
+        cands = [c for c in cands if c['name'] in subtoolchain_names]
+
+        for subtoolchain_name in subtoolchain_names:
+            uniq_subtc_versions = set([subtc['version'] for subtc in cands if subtc['name'] == subtoolchain_name])
+
+            if len(uniq_subtc_versions) == 1:
+                subtoolchain_version = list(uniq_subtc_versions)[0]
+
+            elif len(uniq_subtc_versions) == 0:
+                # only retain GCCcore as subtoolchain if version was found
+                if subtoolchain_name == GCCcore.NAME:
+                    _log.info("No version found for %s; assuming legacy toolchain and skipping it as subtoolchain.",
+                              subtoolchain_name)
+                    subtoolchain_name = GCCcore.SUBTOOLCHAIN
+                    subtoolchain_version = ''
+                # dummy toolchain: end of the line
+                elif subtoolchain_name == DUMMY_TOOLCHAIN_NAME:
+                    subtoolchain_version = ''
+                elif ((subtoolchain_name in optional_toolchains or current_tc_name in optional_toolchains) and
+                      robot_find_easyconfig(subtoolchain_name, current_tc_version)):
+                    # special case: optionally find e.g. golf/1.4.10 for goolf/1.4.10 even if it is not in
+                    # the module dependencies. This is only allowed for and inside optional subtoolchains.
+                    subtoolchain_version = current_tc_version
+                elif subtoolchain_name in optional_toolchains:
+                    continue
+                else:
+                    raise EasyBuildError("No version found for subtoolchain %s in dependencies of %s",
+                                         subtoolchain_name, current_tc_name)
             else:
-                raise EasyBuildError("Multiple versions of %s found in dependencies of toolchain %s: %s",
-                                     subtoolchain_name, current_tc_name, ', '.join(sorted(uniq_subtc_versions)))
+                if subtoolchain_name == DUMMY_TOOLCHAIN_NAME:
+                    # Don't care about multiple versions of dummy
+                    _log.info("Ignoring multiple versions of %s in toolchain hierarchy", DUMMY_TOOLCHAIN_NAME)
+                    subtoolchain_version = ''
+                else:
+                    raise EasyBuildError("Multiple versions of %s found in dependencies of toolchain %s: %s",
+                                         subtoolchain_name, current_tc_name, ', '.join(sorted(uniq_subtc_versions)))
 
-        if subtoolchain_name == DUMMY_TOOLCHAIN_NAME and not build_option('add_dummy_to_minimal_toolchains'):
-            # we're done
-            break
+            if subtoolchain_name == DUMMY_TOOLCHAIN_NAME and not build_option('add_dummy_to_minimal_toolchains'):
+                # we're done
+                continue
```

Comment:
the combination of `we're done` and `continue` here is funny, please change the comment to something more sensible :)